table/tables/mutation_checker.go

```diff
@@ -0,0 +1,262 @@
+// 		the mutations, thus ignored.
+func CheckIndexConsistency(sessVars *variable.SessionVars, t *TableCommon,
+	rowToInsert, rowToRemove []types.Datum, memBuffer kv.MemBuffer, sh kv.StagingHandle) error {
+	if t.Meta().GetPartitionInfo() != nil {
+		return nil
+	}
+	sc := sessVars.StmtCtx
+	if sh == 0 {
+		// some implementations of MemBuffer doesn't support staging, e.g. that in br/pkg/lightning/backend/kv
+		return nil
+	}
+	indexMutations, rowInsertion, err := collectTableMutationsFromBufferStage(t, memBuffer, sh)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if rowToInsert != nil {
+		if err := checkRowInsertionConsistency(sessVars, getOrBuildColumnMap(sessVars.StmtCtx, t), rowToInsert, rowInsertion); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	if err := checkIndexKeys(sc, sessVars, t, rowToInsert, rowToRemove, indexMutations); err != nil {
+		return errors.Trace(err)
+	}
+	// TODO: check whether handles match in index and row mutations
+	return nil
+}
+
+// getOrBuildColumnMap tries to get the column map from stmt ctx. If there isn't one, it builds one and stores it.
+// It saves redundant computations of the map.
+func getOrBuildColumnMap(sc *stmtctx.StatementContext, t *TableCommon) map[int64]*model.ColumnInfo {
+	if sc.ColumnMap == nil {
+		sc.ColumnMap = make(map[int64]map[int64]*model.ColumnInfo)
+	}
+	columnMap, ok := sc.ColumnMap[t.tableID]
+	if !ok {
+		columnMap = make(map[int64]*model.ColumnInfo)
+		for _, col := range t.Meta().Columns {
+			columnMap[col.ID] = col
+		}
+		sc.ColumnMap[t.tableID] = columnMap
+	}
+	return columnMap
+}
+
+// checkIndexKeys checks whether the decoded data from keys of index mutations are consistent with the expected ones.
+func checkIndexKeys(sc *stmtctx.StatementContext, sessVars *variable.SessionVars, t *TableCommon,
+	dataAdded []types.Datum, dataRemoved []types.Datum, indexMutations []mutation) error {
+	indexIDMap := make(map[int64]indexHelperInfo)
+	for _, index := range t.indices {
+		if index.Meta().Primary && t.meta.IsCommonHandle {
+			continue
+		}
+		indexIDMap[index.Meta().ID] = indexHelperInfo{
+			index.Meta(),
+			BuildRowcodecColInfoForIndexColumns(index.Meta(), t.Meta()),
+		}
+	}
+
+	for _, m := range indexMutations {
+		_, indexID, _, err := tablecodec.DecodeIndexKey(m.key)
+		if err != nil {
+			//FIXME: continue or return error?
+			return errors.Trace(err)
+		}
+
+		indexHelperInfo, ok := indexIDMap[indexID]
+		if !ok {
+			return errors.New("index not found")
+		}
+
+		// when we cannot decode the key to get the original value
+		if len(m.value) == 0 && NeedRestoredData(indexHelperInfo.indexInfo.Columns, t.Meta().Columns) {
+			continue
+		}
+
+		decodedIndexValues, err := tablecodec.DecodeIndexKV(m.key, m.value, len(indexHelperInfo.indexInfo.Columns),
+			tablecodec.HandleNotNeeded, indexHelperInfo.rowColInfos)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		indexData := make([]types.Datum, 0)
```

Comment:
And also consider move the declaration of the slice out of the loop started at L120, so that you can reuse it when possible during each loop (I guess the go compiler is not clever enough to automatically optimize this 🤔)